js/packages/openinference-instrumentation-langchain/src/utils.ts

```diff
@@ -0,0 +1,299 @@
+import { Attributes, diag } from "@opentelemetry/api";
+import {
+  assertUnreachable,
+  isNonEmptyArray,
+  isObject,
+  isString,
+} from "./typeUtils";
+import { isAttributeValue } from "@opentelemetry/core";
+import {
+  MimeType,
+  OpenInferenceSpanKind,
+  SemanticConventions,
+} from "@arizeai/openinference-semantic-conventions";
+import { Run } from "@langchain/core/tracers/base";
+import {
+  GenericFunction,
+  LLMMessage,
+  LLMMessageFunctionCall,
+  LLMMessageToolCalls,
+  LLMMessagesAttributes,
+  SafeFunction,
+} from "./types";
+
+export function withSafety<T extends GenericFunction>(fn: T): SafeFunction<T> {
+  return (...args) => {
+    try {
+      return fn(...args);
+    } catch (error) {
+      diag.error(`Failed to get attributes for span: ${error}`);
+      return null;
+    }
+  };
+}
+
+/**
+ * Flattens nested attributes into a single level object
+ */
+function flattenAttributes(
+  attributes: Record<string, unknown>,
+  baseKey: string = "",
+): Attributes {
+  const result: Attributes = {};
+  for (const key in attributes) {
+    const newKey = baseKey ? `${baseKey}.${key}` : key;
+    const value = attributes[key];
+
+    if (value == null) {
+      continue;
+    }
+
+    if (isObject(value)) {
+      Object.assign(result, flattenAttributes(value, newKey));
```

Comment:
just a note that this is modifying the result here. It looks safe but might want a unit test to make sure you are not modifying anything.

Reply:
ahh good point, will spread just to be safe

Reply:
I think this is what we want, result is created at the top of this funciton and returned at the bottom, we want to be modifying it directly as we go long

Reply:
k